metagraph/src/common/utils/file_utils.cpp

```diff
@@ -19,24 +19,27 @@ namespace utils {
 using mtg::common::logger;
 
 std::vector<std::string> TMP_DIRS;
+std::mutex TMP_DIRS_MUTEX;
 
 void cleanup_tmp_dir_on_signal(int sig) {
     logger->trace("Got signal {}. Exiting...", sig);
     // call std::exit to invoke handlers registered in std::atexit
     std::exit(sig);
 }
 
-void cleanup_tmp_dir_on_exit() {
-    for (const std::string &tmp_dir : TMP_DIRS) {
-        logger->trace("Cleaning up temporary directory {}", tmp_dir);
-        try {
-            std::filesystem::remove_all(tmp_dir);
-        } catch (...) {
-            logger->error("Failed cleaning up temporary directory {}", tmp_dir);
-        }
+void cleanup_temp_dir_nolock(const std::filesystem::path &tmp_dir) {
+    logger->trace("Cleaning up temporary directory {}", tmp_dir);
+    try {
+        std::filesystem::remove_all(tmp_dir);
+    } catch (...) {
+        logger->error("Failed to clean up temporary directory {}", tmp_dir);
```

Comment:
Won't this lead to exit code 0 when there's a filesystem error? This isn't ideal...

Reply:
Maybe within the `catch` block you can store the exception in a `std::exception_ptr` and rethrow it after all cleanup is done.

Reply:
Hm, I don't want to throw exceptions here.
It's called either in a signal callback or as a normal single cleanup call.
In both cases, if it failed to remove the temp dir for some reason, it may still finish the rest without issues.
So, just an error message should be enough for now.

I was more concerned about the mutex and locks, which is not fully thread-safe at the moment, in case of finishing a failed run.
But also, don't know how to make it completely safe. Not sure if acquiring a mutex in a signal callback is a good idea.